File: apps/player/apps/native/Sources/Core/MediaTimeline.swift
import Foundation

struct MediaTimeline: Sendable, Equatable {
    let sourceDuration: Double
    let duration: Double
    let omitted: [Range<Double>]

    init(sourceDuration: Double, duration: Double, omitted: [Range<Double>] = []) throws {
        try Input.position(sourceDuration)
        try Input.position(duration)
        guard omitted.count <= 128 else { throw ClientError.invalidResponse }
        var previous: Double = 0
        var removed: Double = 0
        for range in omitted {
            guard range.lowerBound.isFinite, range.upperBound.isFinite, range.lowerBound >= previous,
                  range.upperBound > range.lowerBound, range.upperBound <= sourceDuration else { throw ClientError.invalidResponse }
            previous = range.upperBound
            removed += range.countingLength
        }
        guard abs(sourceDuration - removed - duration) < 0.01 else { throw ClientError.invalidResponse }
        self.sourceDuration = sourceDuration
        self.duration = duration
        self.omitted = omitted
    }

    init(_ raw: JSONValue) throws {
        let value = try raw.object(allowing: ["sourceDuration", "duration", "omitted"])
        let source = try value.requiredNumber("sourceDuration")
        let duration = try value.requiredNumber("duration")
        let ranges = try value.list("omitted", max: 128).map { raw in
            let range = try raw.object(allowing: ["start", "end"])
            let start = try range.number("start")
            let end = try range.requiredNumber("end")
            guard end > start else { throw ClientError.invalidResponse }
            return start..<end
        }
        try self.init(sourceDuration: source, duration: duration, omitted: ranges)
    }

    func sourceTime(_ position: Double) -> Double {
        guard position.isFinite else { return 0 }
        let position = min(duration, max(0, position))
        var removed: Double = 0
        for range in omitted {
            if position < range.lowerBound - removed { break }
            removed += range.countingLength
        }
        return min(sourceDuration, position + removed)
    }

    func presentationTime(_ position: Double) -> Double {
        guard position.isFinite else { return 0 }
        let position = min(sourceDuration, max(0, position))
        var removed: Double = 0
        for range in omitted {
            if position < range.lowerBound { break }
            if position < range.upperBound { return range.lowerBound - removed }
            removed += range.countingLength
        }
        return max(0, position - removed)
    }

    var json: JSONValue {
        .object(["sourceDuration": .number(sourceDuration), "duration": .number(duration),
                 "omitted": .array(omitted.map { .object(["start": .number($0.lowerBound), "end": .number($0.upperBound)]) })])
    }
}

private extension Range where Bound == Double {
    var countingLength: Double { upperBound - lowerBound }
}

File: apps/player/apps/android/app/src/main/java/com/kinosail/player/core/MediaTimeline.kt
package com.kinosail.player.core

import kotlin.math.abs
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull

data class OmittedRange(val start: Double, val end: Double)

class MediaTimeline(val sourceDuration: Double, val duration: Double,
                    val omitted: List<OmittedRange> = emptyList()) {
    init {
        require(sourceDuration.isFinite() && sourceDuration in 0.0..31_536_000.0 &&
            duration.isFinite() && duration in 0.0..31_536_000.0 && omitted.size <= 128) {
            "Invalid playback timeline."
        }
        var last = 0.0
        var removed = 0.0
        omitted.forEach {
            require(it.start.isFinite() && it.end.isFinite() && it.start >= last &&
                it.end > it.start && it.end <= sourceDuration) { "Invalid playback timeline." }
            last = it.end
            removed += it.end - it.start
        }
        require(abs(sourceDuration - removed - duration) < 0.01) { "Invalid playback timeline." }
    }

    fun sourceTime(position: Double): Double {
        val safe = position.coerceIn(0.0, duration)
        var removed = 0.0
        omitted.forEach {
            if (safe < it.start - removed) return (safe + removed).coerceAtMost(sourceDuration)
            removed += it.end - it.start
        }
        return (safe + removed).coerceAtMost(sourceDuration)
    }

    fun presentationTime(position: Double): Double {
        val safe = position.coerceIn(0.0, sourceDuration)
        var removed = 0.0
        omitted.forEach {
            if (safe < it.start) return safe - removed
            if (safe < it.end) return it.start - removed
            removed += it.end - it.start
        }
        return (safe - removed).coerceAtLeast(0.0)
    }

    companion object {
        fun parse(raw: JsonElement): MediaTimeline {
            val value = raw as? JsonObject
            require(value != null && value.keys.all { it in setOf("sourceDuration", "duration", "omitted") } &&
                value.keys.containsAll(setOf("sourceDuration", "duration"))) { "Invalid playback timeline." }
            fun number(rawNumber: JsonElement?): Double {
                val primitive = rawNumber as? JsonPrimitive
                require(primitive != null && !primitive.isString) { "Invalid playback timeline." }
                return primitive.doubleOrNull ?: throw IllegalArgumentException("Invalid playback timeline.")
            }
            val ranges = value["omitted"]?.let {
                val array = it as? JsonArray
                require(array != null && array.size <= 128) { "Invalid playback timeline." }
                array.map { entry ->
                    val range = entry as? JsonObject
                    require(range != null && range.keys == setOf("start", "end")) { "Invalid playback timeline." }
                    OmittedRange(number(range["start"]), number(range["end"]))
                }
            } ?: emptyList()
            return MediaTimeline(number(value["sourceDuration"]), number(value["duration"]), ranges)
        }
    }
}
